File: sdks/typescript/src/http.ts
import { SwarmyApiError } from './error.js';
import type { Problem } from './models.js';

export type FetchLike = (input: string, init?: RequestInit) => Promise<Response>;

export interface SwarmyClientOptions {
  /** Base endpoint of the swarmy controller, e.g. `https://swarm.example.com`. */
  endpoint: string;
  /** API key, e.g. `swk_…`. Sent as `Authorization: Bearer <apiKey>`. */
  apiKey: string;
  /** Override the fetch implementation (defaults to global `fetch`). */
  fetch?: FetchLike;
  /** Extra headers merged into every request. */
  headers?: Record<string, string>;
}

const API_PREFIX = '/api/v1';

/**
 * Thin transport over `fetch`. Handles base-URL joining, bearer auth,
 * JSON encoding, and RFC 9457 error mapping. Resource groups build on top.
 */
export class HttpTransport {
  private readonly base: string;
  private readonly apiKey: string;
  private readonly fetchImpl: FetchLike;
  private readonly extraHeaders: Record<string, string>;

  constructor(opts: SwarmyClientOptions) {
    if (!opts.endpoint) throw new Error('SwarmyClient: `endpoint` is required');
    if (!opts.apiKey) throw new Error('SwarmyClient: `apiKey` is required');
    // Strip trailing slash, then append the versioned API prefix.
    const trimmed = opts.endpoint.replace(/\/+$/, '');
    this.base = trimmed.endsWith(API_PREFIX) ? trimmed : trimmed + API_PREFIX;
    this.apiKey = opts.apiKey;
    const f = opts.fetch ?? (globalThis.fetch as FetchLike | undefined);
    if (!f) throw new Error('SwarmyClient: no fetch implementation available; pass `fetch`');
    this.fetchImpl = f;
    this.extraHeaders = opts.headers ?? {};
  }

  /** Build a full URL for a path (e.g. `/services`) with optional query params. */
  buildUrl(path: string, query?: Record<string, string | number | undefined>): string {
    let url = this.base + path;
    if (query) {
      const params = new URLSearchParams();
      for (const [k, v] of Object.entries(query)) {
        if (v !== undefined) params.set(k, String(v));
      }
      const qs = params.toString();
      if (qs) url += '?' + qs;
    }
    return url;
  }

  async request<T>(
    method: string,
    path: string,
    opts: { query?: Record<string, string | number | undefined>; body?: unknown } = {},
  ): Promise<T> {
    const headers: Record<string, string> = {
      Authorization: `Bearer ${this.apiKey}`,
      Accept: 'application/json, application/problem+json',
      ...this.extraHeaders,
    };
    const init: RequestInit = { method, headers };
    if (opts.body !== undefined) {
      headers['Content-Type'] = 'application/json';
      init.body = JSON.stringify(opts.body);
    }

    const res = await this.fetchImpl(this.buildUrl(path, opts.query), init);

    if (!res.ok) {
      let problem: Problem;
      try {
        problem = (await res.json()) as Problem;
      } catch {
        problem = { type: 'about:blank', title: res.statusText || 'Error', status: res.status };
      }
      throw new SwarmyApiError(res.status, problem);
    }

    // 204 / empty body.
    if (res.status === 204) return undefined as T;
    const text = await res.text();
    if (!text) return undefined as T;
    return JSON.parse(text) as T;
  }
}

File: sdks/python/swarmy/_http.py
"""HTTP transport for the swarmy client (hand-written, stable).

Uses the standard library only (``urllib``). The opener is injectable so it
can be mocked in tests.
"""

from __future__ import annotations

import json
import urllib.error
import urllib.parse
import urllib.request
from typing import Any, Dict, Optional

from .errors import SwarmyApiError
from .models import Problem

API_PREFIX = "/api/v1"


class HttpTransport:
    def __init__(
        self,
        endpoint: str,
        api_key: str,
        opener: Optional[Any] = None,
        timeout: float = 30.0,
        headers: Optional[Dict[str, str]] = None,
    ) -> None:
        if not endpoint:
            raise ValueError("SwarmyClient: `endpoint` is required")
        if not api_key:
            raise ValueError("SwarmyClient: `api_key` is required")
        trimmed = endpoint.rstrip("/")
        self.base = trimmed if trimmed.endswith(API_PREFIX) else trimmed + API_PREFIX
        self.api_key = api_key
        self.timeout = timeout
        self.extra_headers = headers or {}
        # urllib OpenerDirector-like object; defaults to the module-level opener.
        self._opener = opener or urllib.request.build_opener()

    def build_url(self, path: str, query: Optional[Dict[str, Any]] = None) -> str:
        url = self.base + path
        if query:
            pairs = [(k, str(v)) for k, v in query.items() if v is not None]
            if pairs:
                url += "?" + urllib.parse.urlencode(pairs)
        return url

    def request(
        self,
        method: str,
        path: str,
        query: Optional[Dict[str, Any]] = None,
        body: Optional[Any] = None,
    ) -> Any:
        url = self.build_url(path, query)
        data = None
        headers = {
            "Authorization": "Bearer {}".format(self.api_key),
            "Accept": "application/json, application/problem+json",
        }
        headers.update(self.extra_headers)
        if body is not None:
            data = json.dumps(body).encode("utf-8")
            headers["Content-Type"] = "application/json"

        req = urllib.request.Request(url, data=data, headers=headers, method=method)
        try:
            resp = self._opener.open(req, timeout=self.timeout)
        except urllib.error.HTTPError as exc:  # non-2xx
            raw = exc.read()
            problem = self._parse_problem(raw, exc.code, exc.reason)
            raise SwarmyApiError(exc.code, problem) from None

        with resp:
            status = getattr(resp, "status", 200) or 200
            raw = resp.read()
        if not raw:
            return None
        return json.loads(raw.decode("utf-8"))

    @staticmethod
    def _parse_problem(raw: bytes, status: int, reason: Any) -> Problem:
        try:
            return Problem.from_dict(json.loads(raw.decode("utf-8")))
        except Exception:
            return Problem(
                type="about:blank",
                title=str(reason) if reason else "Error",
                status=status,
            )
